core/src/smartcontracts/isi/mod.rs

```diff
@@ -33,6 +36,51 @@ pub trait Registrable {
     fn build(self, authority: &AccountId) -> Self::Target;
 }
 
+/// Validate and execute an additional instruction `T` during `Self` execution.
+/// FIXME This means going backwards from execution phase to validation phase, which is contrary to current design principles
+trait BackValidate<T>
```

Comment:
IMO it is not the responsibility of `iroha_core` to create destination account if it doesn't exist, it is executor's responsibility. I would move all this logic into default executor and `iroha_core` should plainly reject all instructions that operate on non-existing accounts. In other words, default executor should check whether destination account exists and if not first execute `Register<Account>` which will go through the regular validation process (since validation is done in executor) 

If you do this you wouldn't have the problem you are describing

Reply:
@mversic my initial concern was exactly about demanding registration permissions when i.e. making transfer, imo we shouldn't do that

Reply:
I see. IMO, in any case `iroha_core` should reject this transfer if destination doesn't exist. It is up to the executor to create missing destination before doing the transfer. We can make the default executor not require any permissions when creating a provisional account, but it has to create it before executing the transfer

Reply:
I'll try to implement that in the following PR, which will alternate the 2nd commit of this PR.
There, executor would try to provisionally register the account anyway without checking if the account is already registered.
This is because the executor does not refer to the state -- if it needs to determine if the account is existing, the flag should be included in AccountId, which leads to a perversion that the transaction builder should concern about whether the account is registered or not

Reply:
> I see. IMO, in any case `iroha_core` should reject this transfer if destination doesn't exist. It is up to the executor to create missing destination before doing the transfer. We can make the default executor not require any permissions when creating a provisional account, but it has to create it before executing the transfer
> 
> I guess accounts will then have a flag to indicate whether they are provisional or not

If we are going to settle down with the approach that no additional permissions are required to make transfer, i think it's okay to create necessary for this transfer structures inside core itself.

Reply:
> If we are going to settle down with the approach that no additional permissions are required to make transfer, i think it's okay to create necessary for this transfer structures inside core itself.

I don't think so. We can go about this 2 ways:
1. Iroha is maximally permissible (it will create necessary conditions for instruction to succeed)
2. Iroha is maximally restrictive (if conditions are not met it will reject the instructions)

IMO execution in Iroha should have no side effects and we should go with 2nd approach. All side effects should be contained in the executor

> I'll try to implement that in the following PR, which will alternate the 2nd commit of this PR.

all I'm saying is that you should put this logic into executor. You can deal with "provisional" accounts in a separate PR

Reply:
Let's consider that for some extreme case deployment where user using really minimal executor, without additional logic of creating provisional account user can't access feature of offline accounts.
Is it something we want?

Reply:
> This is a kind of compromise because it goes against the current design principle that all validations should be done prior to any execution.

@s8sato this is not always the case, on `Unregister<Role>` we execute unvalidated `Revoke<Role, Account>` for all accounts that have it granted. As was noted in #4330

> Let's consider that for some extreme case deployment where user using really minimal executor, without additional logic of creating provisional account user can't access feature of offline accounts.
Is it something we want?

ok, I have a different suggestion. Let's keep the current logic in `iroha_core` but have our `default_executor` forbid this by default. I say this since I don't find this is something that permissioned blockchains would want to have enabled.

Reply:
ok, it could be the starting point (allowing it later would be backward compatible)

Reply:
@Erigara could you more specific why the `Map<AccountId, Asset>` would achieve an equivalent of provisional accounts?
I understand all we have to do to apply e.g. transfer instruction would be modifying the map only.

I'm not sure it is relevant to the following, but
it reminded me of what I called `@nodomain` approach, which represents being provisional by being under "nodomain".
Why I rejected this approach is because it seemed obviously unreasonable to replace e.g. `@wonderland` with `@nodomain` and remember the original `@wonderland` and later modify `@nodomain` to `@wonderland`.
But an evil here might be non-separation of the account identification and the domain membership.
If they were separated, being under "nodomain" (and being provisional) would have been equal to `account.domain_id: None`.
This approach might rather work well with the shallow objects

Reply:
In that case we still need something indicates if the source account is active or not

Reply:
Yeah, in that case creating Account and activating it happens at the same time

Reply:
If there is no need for provisional registration we can check that source account is registered and it should be enough, shouldn't it?